docs/research/ha/RATIONALE.md

```diff
@@ -0,0 +1,446 @@
+# High Availability with s3gw
+
+- [High Availability with s3gw](#high-availability-with-s3gw)
+  - [Active-Active pipelines](#active-active-pipelines)
+  - [Active-Passive pipelines](#active-passive-pipelines)
+  - [One Active pipeline](#one-active-pipeline)
+  - [Investigation's Rationale](#investigations-rationale)
+  - [Failure cases](#failure-cases)
+    - [radosgw's POD failure](#radosgws-pod-failure)
+    - [Cluster's node failure](#clusters-node-failure)
+    - [radosgw's failure due to a bug not related to a certain input pattern](#radosgws-failure-due-to-a-bug-not-related-to-a-certain-input-pattern)
+    - [radosgw's failure due to a bug related to a certain input pattern](#radosgws-failure-due-to-a-bug-related-to-a-certain-input-pattern)
+    - [PV Data corruption at application level due to radosgw's bug](#pv-data-corruption-at-application-level-due-to-radosgws-bug)
+    - [PV Data corruption at application level due to radosgw's anomalous exit](#pv-data-corruption-at-application-level-due-to-radosgws-anomalous-exit)
+  - [Measuring s3gw failures on Kubernetes](#measuring-s3gw-failures-on-kubernetes)
+  - [The s3gw Probe](#the-s3gw-probe)
+  - [Tested scenarios](#tested-scenarios)
+    - [EXIT-1, 10 measures](#exit-1-10-measures)
+    - [EXIT-0, 10 measures](#exit-0-10-measures)
+    - [Considerations on the tests performed](#considerations-on-the-tests-performed)
+
+We want to investigate what *High Availability* - HA - means for a project like
+the s3gw.
+
+If we identify the meaning of HA as the ability to have N independent pipelines
```

Comment:
HA is not necessarily tied to multiple independent pipelines. While High Availability (as a component of dependable computing) can be achieved through redundancy, it can also be achieved by significantly increased reliability and quality of the stack. e.g., improving the start time improves HA by lowering the Mean Time To Repair (MTTR) metric after detecting and recovering from a fault. And achieving HA in a way that *never* affects the user's operations is also not realistic, since not all faults can be masked, or the complexity required to attempt it would actually be detrimental to overall system reliability/availability (aka not cost-effective). Paradoxically, failing fast instead of hanging on an attempted fault recovery can also increase observed availability.

I'd argue our key goal is to maximize user-visible, reliable service availability, and to be tolerant with regard to certain faults (such as pod crashes, single node outages ... - the faults we consider or explicitly exclude are discussed later in the document).

Reply:
> I'd argue our key goal is to maximize user-visible, reliable service availability, and to be tolerant with regard to certain faults

I Agree on the goal's high level idea; we have to translate this statement in a precise requisite that can be then implemented.
I'd say also that the HA model could be very much dependent to the specific need of the user; for a certain workload could be more important:

- After have experienced a failure, have the guarantee that the S3 service will be available after X seconds. 
- After have experienced a failure, have the guarantee that the S3 service will be available on the NEXT (+X) access. (where X is something that goes from 0..N)

Reply:
Right. Since we don't know how many ops/s the user pushes, and our recovery time is (mostly) constant (possibly related to the number of ops we need to replay from the journal etc?), I think we'd go with the "Mean Time to Repair" or more likely "Recovery Time Objective" in response to the failure.

That also allows us to specify the "Recovery *Point* Objective" - which operations are possibly lost? e.g., in-flight operations at the time of most faults will have to be retried by the client side (we can't resume the TCP/IP connection).

Reply:
With a little more time, it seems that "RTO" is more common in the storage and DR/BC worlds. (MTTR is more often used to repairing the underlying cause.)

It also fits better with our "RPO" concept (which should be zero for completed and acknowledged operations, in-flight and not acknowledged ones might need to be restarted by the client).